lib/Frontend/ParseableInterfaceModuleLoader.cpp

```diff
@@ -0,0 +1,1015 @@
+
+namespace {
+
+/// The result of a search for a module either alongside an interface, in the
+/// module cache, or in the prebuilt module cache.
+class ModuleDiscovery {
+  /// If the discovered module is a forwarding module, this stores the contents
+  /// of the forwarding file.
+  ForwardingModule forwardingModule;
+
+  /// Whether or not we've loaded the forwarding module. It is an error to
+  /// access the forwarding module if this is \c false.
+  bool hasLoadedForwardingModule = false;
+
+  /// The kind of module we've found.
+  enum class Kind {
+    /// A module that's either alongside the swiftinterface or in the
+    /// module cache.
+    Normal,
+
+    /// A module that resides in the prebuilt cache, and has hash-based
+    /// dependencies.
+    Prebuilt,
+
+    /// A 'forwarding' module. \sa ForwardingModule.
+    Forwarding
+  };
+
+  /// The kind of module that's been discovered.
+  Kind kind;
+  ModuleDiscovery(StringRef path, Kind kind): kind(kind), path(path) {}
+public:
+
+  /// The path to the discovered module on disk.
+  std::string path;
+
+  /// Creates a \c Normal discovered module.
+  static ModuleDiscovery normal(StringRef path) {
+    return { path, Kind::Normal };
+  }
+
+  /// Creates a \c Prebuilt discovered module.
+  static ModuleDiscovery prebuilt(StringRef path) {
+    return { path, Kind::Prebuilt };
+  }
+
+  /// Creates a \c Forwarding discovered module.
+  static ModuleDiscovery forwarding(StringRef path) {
+    return { path, Kind::Forwarding };
+  }
+
+  bool isNormal() const { return kind == Kind::Normal; }
+  bool isPrebuilt() const { return kind == Kind::Prebuilt; }
+  bool isForwarding() const { return kind == Kind::Forwarding; }
+
+  /// Loads the underlying \c Forwarding module.
+  /// \note: This must be called before calling \c getForwardingModule().
+  bool loadForwardingModule(llvm::vfs::FileSystem &fs) {
+    auto fwdBuf = fs.getBufferForFile(path);
+    if (!fwdBuf) return false;
+    llvm::yaml::Input yamlIn(fwdBuf.get()->getBuffer());
+    yamlIn >> forwardingModule;
+    hasLoadedForwardingModule = true;
+    return true;
+  }
+
+  const ForwardingModule &getForwardingModule() const {
+    assert(kind == Kind::Forwarding &&
+           "getForwardingModule() called on non-forwarding module");
+    assert(hasLoadedForwardingModule &&
+           "must loading forwarding module contents before accessing");
+    return forwardingModule;
+  }
+
+  // Returns the path to the .swiftmodule that will eventually be loaded by
+  // this
+  StringRef getModulePath() const {
+    switch (kind) {
+      case Kind::Normal:
+      case Kind::Prebuilt:
+        return path;
+      case Kind::Forwarding:
+        return getForwardingModule().underlyingModulePath;
+    }
+  }
+};
+
+} // end anonymous namespace
+
+#pragma mark - Common utilities
+
+namespace path = llvm::sys::path;
+
+static std::unique_ptr<llvm::MemoryBuffer>
+getASTBuffer(llvm::vfs::FileSystem &fs, StringRef path) {
```

Comment:
Nitpick: can you call this "SerializedAST" instead of just "AST"? I had to think about what this meant.